### ISSUE TITLE
Diagnosing multi-threading segfaults upon upgrading to 3.1.9

### ISSUE THREAD

**Issue by LTLA** [emscripten-core/emscripten#16827]
**tl;dr** Updating to 3.1.9 resulted in the introduction of a sporadic segfault in my multi-threaded code.

**Minimal reproducible example**: https://github.com/LTLA/h5wasm-pthreads-MRE

**Long story:**

My application links against an Emscripten build of the HDF5 library (see https://github.com/usnistgov/libhdf5-wasm). It also uses Emscripten's pthreading support for parallelization. So far, both of these features have happily co-existed in the same Wasm binary... until 3.1.9, at which point the app began to experience a variety of strange and sporadic run-time errors related to the threading.

The MRE repository contains the boiled-down version of the problem. In short, linking to the HDF5 static library is sufficient to cause rare failures in the threaded code, even though the HDF5 functions are never called (in either the parallel or serial regions). In fact, in the MRE, the threads don't do anything at all; the segfaults occur during the creation of the thread, looking like this:

```
RuntimeError: Aborted(segmentation fault)
    at abort (file:///Users/luna/Programming/js/h5-mre/build/minimal.js:1506:10)
    at segfault (file:///Users/luna/Programming/js/h5-mre/build/minimal.js:924:2)
    at SAFE_HEAP_LOAD_i32_1_U_1 (<anonymous>:wasm-function[605]:0x10071)
    at _emscripten_thread_sync_code (<anonymous>:wasm-function[226]:0x51ea)
    at _emscripten_yield (<anonymous>:wasm-function[241]:0x5a52)
    at emscripten_futex_wait (<anonymous>:wasm-function[227]:0x52cd)
    at __timedwait_cp (<anonymous>:wasm-function[284]:0x7291)
    at __timedwait (<anonymous>:wasm-function[285]:0x72f8)
    at __pthread_mutex_timedlock (<anonymous>:wasm-function[310]:0x7d74)
    at __pthread_mutex_lock (<anonymous>:wasm-function[308]:0x7c62)
```

Now, I don't think this is completely Emscripten's fault, because the HDF5 library is notoriously finicky w.r.t. any parallelization. My current hypothesis about the nature of this bug (repeated from usnistgov/libhdf5-wasm#1) is that:

1. The HDF5 library is setting some global state when it starts up.
2. When a Emscripten-managed web worker is asked to execute something for the first time, it runs this start-up code. (It seems that the actual initialization of the worker doesn't do all of the Wasm set-up, if my interpretation of the `*.worker.js` file is correct.)
3. If you have enough workers that are launched at the same time, all of them are fighting with each other to perform the global set-up. This causes some kind of corruption in the SharedArrayBuffer, resulting in a segfault.

This leaves me with two questions:

- Is my hypothesis correct? I'd like to figure out what Emscripten included from the HDF5 static library to see what exactly what it was doing in the Wasm binary, but unfortunately my wat-reading skills are not up to scratch; any help would be appreciated.
- Why does this occur in 3.1.9 and not in 3.1.8? I see some changes related to proxying to threads but I don't know if that's relevant.

My current workaround is to wait for thread set-up to complete (by polling on some atomic flags) before creating the next thread. This seems to work but I would be a lot more confident in it if I understood why.

**Comment by LTLA** [comment]
Thanks @sbc100; I was just doing the bisection as suggested by @kripken, and I reached the same conclusion.  `c8f9aeea265a89f87787da70a35cbf29b9dac7bf` is the release that introduces this behavior.

> Can we see the full stacktrace? 

There's not much more than what I wrote above, but here's the full message that Node spits out (built using `c8f9aeea265a89f87787da70a35cbf29b9dac7bf`; the message in the original post is built with 3.1.9).

```
Aborted(segmentation fault)
worker.js onmessage() captured an uncaught exception: RuntimeError: Aborted(segmentation fault)
RuntimeError: Aborted(segmentation fault)
    at abort (file:///Users/luna/Programming/js/h5-mre/build/minimal.js:1498:10)
    at segfault (file:///Users/luna/Programming/js/h5-mre/build/minimal.js:916:2)
    at SAFE_HEAP_LOAD_i32_1_U_1 (<anonymous>:wasm-function[623]:0x103f9)
    at _emscripten_thread_sync_code (<anonymous>:wasm-function[241]:0x5665)
    at _emscripten_yield (<anonymous>:wasm-function[256]:0x5ecd)
    at emscripten_futex_wait (<anonymous>:wasm-function[242]:0x5748)
    at __timedwait_cp (<anonymous>:wasm-function[299]:0x75b6)
    at __timedwait (<anonymous>:wasm-function[300]:0x761d)
    at __pthread_mutex_timedlock (<anonymous>:wasm-function[325]:0x8097)
    at __pthread_mutex_lock (<anonymous>:wasm-function[323]:0x7f85)

file:///Users/luna/Programming/js/h5-mre/build/minimal.js:308
   throw ex;
   ^
```

If there's any more debugging information, I'm happy to stick in the relevant flags.

> Do you know on which thread the segfault occurs?

No, but I do observe it on multiple threads on occasion. I can try to pull out some more information about the ID of the failing thread(s).

> Are you building with dynamic linking (-sMAIN_MODULE)? Are you loading side modules?

Not that I'm aware of.

**Comment by sbc100** [comment]
Ah, can you run node with `--stack-trace-limit=50` to get a bigger stack?

**Comment by sbc100** [comment]
Thanks!  That backtrace makes it very clear what the problem is.   `_emscripten_thread_sync_code` code uses TLS but its being called from within the tls init code.    This will happen if there is contension from the malloc lock while the thread is starting up.  Certainly looks like a real emsripten bug to me.

**Comment by chenzx** [comment]
https://github.com/emscripten-core/emscripten/issues/16968 seems related... (my legacy C++ code base uses thread_local globals and starts 10 js worker threads and also got a lot of similar aborts)

**Comment by sbc100** [comment]
The problem is _emscripten_thread_sync_code being called before TLS is done initializing.   I'm working on a fix.

**Comment by shimaode** [comment]
Hi, Any updates on this issue?  I'm encountering the same problem with 3.1.13, trace stack is as follows:
```
Uncaught RuntimeError: Aborted(segmentation fault)
    at abort (client.js:1250:10)
    at segfault (client.js:649:2)
    at SAFE_HEAP_LOAD_i32_1_U_1 (emscripten_thread_init.c:25:1)
    at _emscripten_thread_sync_code (dynlink.c:226:7)
    at _emscripten_yield (emscripten_yield.c:45:3)
    at emscripten_futex_wait (emscripten_futex_wait.c:116:3)
    at __timedwait_cp (__timedwait.c:93:8)
    at __timedwait (__timedwait.c:116:6)
    at __pthread_mutex_timedlock (pthread_mutex_timedlock.c:89:7)
    at __pthread_mutex_lock (pthread_mutex_lock.c:10:1)
```
Any workaround or suggestions would be appreciated.

**Comment by sbc100** [comment]
I believe this was fixed in #17090.   Please re-open if you continue to see this issue.
